src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +407,60 @@ where
                                     // We store errors as well as non-None values, so that they can be
                                     // retracted if new rows show up for the same key.
                                     let new_value = decoded_value.transpose();
-                                    let old_value = if let Some(new_value) = &new_value {
-                                        current_values.insert(decoded_key, new_value.clone())
-                                    } else {
-                                        current_values.remove(&decoded_key)
+
+                                    let (thinned_new_value, thinned_old_value) =
+                                        if let Some(new_value) = new_value {
+                                            let thinned_value = new_value.map(|full_row| {
+                                                row_packer.clear();
+                                                /*
+                                                // Many parts of the codebase rely on indexing,
+                                                // so we have to fill in the empty columns with
+                                                // dummy data
+                                                row_packer.extend(
+                                                    (0..key_arity).map(|_| Datum::Dummy),
+                                                );*/
+                                                row_packer.extend(full_row.iter().skip(key_arity));
+                                                row_packer.finish_and_reuse()
+                                            });
+                                            (
+                                                Some(thinned_value.clone()),
+                                                current_values.insert(
+                                                    decoded_key.clone(),
+                                                    // Thin out the row to not contain a copy of the key(s)
+                                                    thinned_value,
+                                                ),
+                                            )
+                                        } else {
+                                            (None, current_values.remove(&decoded_key))
+                                        };
+
+                                    // This closure re-uses a `Row` to
+                                    // rebuild the full `Row` with both keys and values, before we
+                                    // send them off to sql-land
```

Comment:
It's not really precise to say we are doing this before sending the rows to SQL (even though that's technically true, but it happens like... way before). It would make more sense to say we are rebuilding the row before outputting it from this operator.